decepticonlp/attacks/attack.py

```diff
@@ -0,0 +1,328 @@
+	"""
+
+    def __init__(
+        self,
+        model,
+        data_loader,
+        adversarial_data_loader,
+        input_format,
+        accuracy=True,
+        logs_after_every=50,
+        device=torch.device("cpu"),
+    ):
+
+        # initialise parameters
+        self.model = model.to(device)
+        self.data_loader = data_loader
+        self.adversarial_data_loader = adversarial_data_loader
+        self.input_format = input_format
+        self.accuracy = accuracy
+        self.logs_after_every = logs_after_every
+        self.device = device
+
+        self.loss_logs = {}
+        self.accuracy_logs = {}
+
+    def attack(self):
+
+        # set model to eval mode
+        self.model.eval()
+
+        print("Running inference on original dataset")
+        (
+            self.loss_logs["original"],
+            self.accuracy_logs["original"],
+        ) = self.inference_loop(self.data_loader)
+
+        print("Running inference on adversarial dataset")
+        (
+            self.loss_logs["adversarial"],
+            self.accuracy_logs["adversarial"],
+        ) = self.inference_loop(self.adversarial_data_loader)
+
+    def inference_loop(self, dataloader):
+        # Define inference loop for dataloader
+        running_loss = 0.0
+        correct_total = 0
+        counter = 1
+        loss_logs = []
+        accuracy_logs = []
+        tqdm_dataloader = tqdm(dataloader)
+
+        for batch in tqdm_dataloader:
+            inputs = batch
+            input_dict = {}
+            for k, i in enumerate(self.input_format):
+                input_dict[i] = inputs[k].long().to(self.device)
+
+            # Forward pass
+            outputs = self.model(**input_dict)
+            # Compute Loss and probabilities
+            computed_loss = outputs[0]
+            logits = outputs[1]
+            probabilities = torch.nn.functional.log_softmax(logits, dim=1)
+
+            running_loss += computed_loss.item()
+            loss = running_loss / (counter * dataloader.batch_size)
+
+            # Log the loss
+            if counter % self.logs_after_every == 0:
+                loss_logs.append(loss)
+
+            tqdm_dataloader.set_description("Loss (%f)" % (loss))
+
+            # Compute Accuracy
+            if self.accuracy:
+                _, computed_labels = torch.max(probabilities, dim=1)
+                correct = (computed_labels == input_dict["labels"]).float().sum()
+
+                correct_total += correct
+                acc = correct_total / (counter * dataloader.batch_size)
+
+                if counter % self.logs_after_every == 0:
+                    accuracy_logs.append(acc)
+
+                tqdm_dataloader.set_postfix(Accuracy=acc)
+
+            # Increment counter
+            counter += 1
+        return loss_logs, accuracy_logs
+
+    # Getter functions
+
+    def get_criterion_logs(self):
+        return self.loss_logs, self.accuracy_logs
+
+
+class CharAttacker:
+    """
```

Comment:
Need to add a description of what the class does

Reply:
On it